core/lib/multivm/src/versions/vm_fast/validation_tracer.rs

```diff
@@ -0,0 +1,291 @@
+use std::collections::HashSet;
+
+use zk_evm_1_3_1::address_to_u256;
+use zksync_types::{
+    u256_to_address, Address, ACCOUNT_CODE_STORAGE_ADDRESS, BOOTLOADER_ADDRESS,
+    CONTRACT_DEPLOYER_ADDRESS, KECCAK256_PRECOMPILE_ADDRESS, L2_BASE_TOKEN_ADDRESS,
+    MSG_VALUE_SIMULATOR_ADDRESS, SYSTEM_CONTEXT_ADDRESS, U256,
+};
+use zksync_vm2::interface::{
+    CallframeInterface, GlobalStateInterface, Opcode::*, OpcodeType, ReturnType::*, ShouldStop,
+    Tracer,
+};
+use zksync_vm_interface::tracer::{
+    TimestampAsserterParams, ValidationParams, ValidationTraces, ViolatedValidationRule,
+};
+
+use super::utils::read_fat_pointer;
+use crate::tracers::TIMESTAMP_ASSERTER_FUNCTION_SELECTOR;
+
+pub trait ValidationMode: Tracer + Default {
```

Comment:
Bikeshedding: IMO, `ValidatingTracer` would be a better name (or maybe `ValidationTracer` + rename `ValidationTracer` -> `FullValidationTracer`). As it's named right now, I'd not expect `Tracer` to be a supertrait, but rather have it as an associated type.